python/ql/src/Security/CWE-327/InsecureProtocol.ql

```diff
@@ -0,0 +1,60 @@
+/**
+ * @name Use of insecure SSL/TLS version
+ * @description An insecure version of SSL/TLS has been specified. This may
+ *              leave the connection open to attacks.
+ * @id py/insecure-protocol
+ * @kind problem
+ * @problem.severity warning
+ * @precision high
+ * @tags security
+ *       external/cwe/cwe-327
+ */
+
+import python
+
+FunctionObject ssl_wrap_socket() {
+    result = any(ModuleObject ssl | ssl.getName() = "ssl").getAttribute("wrap_socket")
+}
+
+string insecure_version_name() {
+    // For the `ssl` module
+    result = "SSLv2_METHOD" or
+    result = "SSLv23_METHOD" or
+    result = "SSLv3_METHOD" or
+    result = "TLSv1_METHOD" or
+    // For `pyOpenSSL.SSL`
+    result = "PROTOCOL_SSLv2" or
+    result = "PROTOCOL_SSLv3" or
+    result = "PROTOCOL_TLSv1"
+}
+
+private ModuleObject the_ssl_module() {
+    result = any(ModuleObject m | m.getName() = "ssl")
+}
+
+private ModuleObject the_pyOpenSSL_module() {
+    result = any(ModuleObject m | m.getName() = "pyOpenSSL").getAttribute("SSL")
+}
+
+predicate unsafe_ssl_wrap_method_call(CallNode call) {
+    call = ssl_wrap_socket().getACall() and
+    exists(ControlFlowNode arg | arg = call.getArgByName("ssl_version") |
+        arg.(AttrNode).getObject(insecure_version_name()).refersTo(the_ssl_module())
```

Comment:
`arg.refersTo(the_ssl_module().getAttribute(insecure_version_name())` would be better.

Reply:
For this and the above comment, I'm not sure that'll work unless I also stub the `ssl` module. It seems many of the old insecure protocols have been deprecated (and for good reason), hence the use of a more "syntactic" check. I can add the `refersTo` in addition to the syntactic check, however.

Reply:
Are the "METHOD" constants the correct names. For Python 2.7.12
```
>>> [ n for n in dir(ssl) if 'METHOD' in n ]
[]
```

Reply:
`METHOD` is for `pyOpenSSL`.
```ql
>>> import ssl
>>> [n for n in dir(ssl) if 'PROTOCOL' in n]
['ALERT_DESCRIPTION_PROTOCOL_VERSION', 'PROTOCOL_SSLv23', 'PROTOCOL_TLS', 'PROTOCOL_TLSv1', 'PROTOCOL_TLSv1_1', 'PROTOCOL_TLSv1_2', '_PROTOCOL_NAMES']
```

Reply:
In which case the comments in `insecure_version_name()` are the wrong way around.